pumount.1
.TH pumount 1 "August 27, 2004" "Martin Pitt"

.SH NAME
pumount \- umount arbitrary hotpluggable devices as normal user

.SH SYNOPSIS

.B pumount
[
.I options
]
.I device

.SH DESCRIPTION

pumount is a wrapper around the standard umount program which permits normal
users to umount removable devices without a matching /etc/fstab entry. 

pumount also supports encrypted devices which use dm-crypt and have
LUKS metadata. If a LUKS-capable
.B cryptsetup
is installed, pumount will umount the mapped device instead and call
cryptsetup to close the decrypted device afterwards.

.B pumount
expects the
.I device
as its only argument. This will umount 
.I device
from a directory below /media if policy is met (see below). 

.SH OPTIONS

.TP
.B \-l, \-\-lazy
Lazy unmount. Detach the filesystem from the filesystem hierarchy now,
and cleanup all references to the filesystem as soon as it is not busy
anymore.  (Requires kernel 2.4.11 or later.)

.TP
.B \-h, \-\-help
Print a help message and exit successfully.

.TP
.B \-d, \-\-debug
Enable verbose debug messages.

.SH POLICY

The umount will succeed if all of the following conditions are met:

.IP \(bu 2
.I device
is a block device in /dev/ (it does not need to exist if
.B \-l
is supplied)
.IP \(bu 
.I device
is not in /etc/fstab (if it is, pmount executes \fB umount \fI device\fR as the
calling user to handle this transparently)
.IP \(bu
.I device
is mounted according to /etc/mtab and /proc/mounts with the calling user's uid
.IP \(bu
mount point is in /media

.SH SEE ALSO

.BR pmount (1),
.BR umount (8)

.SH AUTHOR
.B pmount
is developed by Martin Pitt <martin.pitt@canonical.com>.


CHANGES
0.9.2
-----
- Switched version control to bazaar-ng (http://www.bazaar-ng.org);
  publicly accessible branch is at
  http://people.ubuntu.com/~pitti/bzr/pmount/
- pmount-hal: ignore invalid file systems, autodetect in this case
- Mark JFS and iso9660 as iocharset-capable
- pmount-hal: read volume.policy.mount_option.umask from hal.
- Fix "make dist" to work properly with the bazaar-ng metadata

0.9.1
-----
- Makefile: fix "make dist" tarball generation
- pmount: use async by default to not destroy flash memory

0.9
---
- pmount-hal: ported to new hal 0.5 and dbus 0.32 API
- integrate cryptsetup (with LUKS extension) if device is encrypted
  with dmcrypt and LUKS metadata
- pmount: mount vfat with shortname=winnt option
- pmount-hal: use volume.policy.mount_filesystem instead of
  volume.fstype
- always build the POT file, delete it on "make clean"
- Makefile: added "dist" target

0.8
---
- unbreak locking, this created ".created_by_pmount" stamps in the
  locking directories
- accept pumount by label name
- accept pmount by mount path (only in /media)
- added --umask option
- renamed template.pot to pmount.pot to match translation domain
- support a whitelist in /etc/pmount.allow

0.7.2
-----
- updated translations from Rosetta: ca, es_ES
- added new translations from Rosetta: cs, es, fr, hr, it_IT, nb

0.7.1
-----
- pmount-hal: consider a directory empty if it only contains a
  .created-by-pmount file; this makes it work with unmounting using
  umount instead of pumount

0.7
---
- support --exec/-e option
- rewrote pmount-hal in C, now it runs 10 times faster

0.6
---
- pmount-hal: use hal-supplied file system type (if present) to speed up
  mounting and to avoid bogus error messages in kernel log
- mount vfat with 'quiet' option
- support pmount/pumount'ing mount points
- added Catalan translation, thanks to Eduard Duran
- added Spanish translation, thanks to Gustavo de Lama and Jorge
  González González

0.5.1
-----
- pmount: do not allow empty media labels; this allowed to mount a
  device directly under /media if /media was empty

0.5
---
- use stamp files to mark directories created by pmount and only
  remove a mount point directory if it contains a stamp file
- use the given device name as default mount point name instead of the
  symlink-resolved one
- i18n support
- added German translation

0.4.4
-----
- pmount-hal: resolve symlinks before checking mount point directories
- use umask "077" instead of "007" for VFAT and NTFS volumes; this allows e. g.
  to use ssh keys on USB sticks
- if a mount with given iocharset fails, try again without; there were reports
  of failed vfat mounts when giving iocharset

0.4.3
-----
- pmount-hal: when checking for already existing mount points, also
  accept existing, but empty directories (if nothing is mounted at
  them)

0.4.2
-----
- pumount: allow to unmount any device if called as root

0.4.1
-----
- consistently use the realpath() of devices for all operations but
  transparent mount fallback

0.4
---
- pmount: added --noatime option
- Makefile: support PREFIX
- Makefile: install manapges, too
- Makefile: add uninstall target
- added pmount-hal (incl. manpage)

0.3
---
- pmount: added option -a/--async for asynchronous mounting
- pmount: added option -t for selecting a file system
- added NTFS support
- pmount: added option -c/--charset for selecting the iocharset;
  default to 'utf8' when pmount is executed in an UTF-8 locale,
  otherwise default to mount default
- pmount: do not fail if a device is handled in /etc/fstab, but a
  label is supplied; merely print a warning

0.2.2
-----
- mount CD-ROM file systems without umask parameter (this does not work)
- added and documented dynamic debug option -d, use dynamic debug output
  everywhere
- pumount: now use getopt_long() for option parsing, added -d and -h

0.2.1
-----
- version 0.2 was not published, had a major policy hole: in some cases it
  allowed to mount partitions on fixed disks
- policy.c: rewrote find_sysfs_device() to check minor device number, too

0.2 [DO NOT USE]
----------------
- major refactorization to clean up the code
- consolidated error return codes
- pmount: added --lock/--unlock commands
- device is regarded as removable also if /sys/block/drive/removable is 1
- mount devices with umask=007

0.1
---
- fixed /dev/ path checking: now use the real path of the device node, not the
  path supplied on the command line
- pmount: don't try to mount all file systems with the uid and gid options,
  this does not work on ext3, ext2, reiserfs, xfs, and jfs; try two separate
  mount commands now
- pumount: added option -l to support lazy unmounting
- pumount: changed policy: 
  * device must be mounted by the same uid that attempts to pumount
  * device node does not need to exist and be removeable any more (because the
    device node cannot be checked any more if device was already ripped out)

0.0.8
-----
- use root privileges also for checking that the existing mount point is empty,
  since normal user might not be allowed to read the directory

0.0.7
-----
- replaced all str[n]cat() and str[n]cpy() with snprintf() calls; thanks to
  Matt Zimmerman for the patch

0.0.6
-----
- added support for CD-ROM file systems (udf and iso9660)
- require mount points to be empty before mount

0.0.5
-----
- now also handle devices in /etc/fstab, transparently call mount/umount in
  this case

0.0.4
-----
- drop effective user id to normal user until root is really needed
- added proper online help
- added manpages

0.0.3
-----
- pumount: added similar suid check 
- pumount: delete mount point after umounting
- determine real path of devices before checking them
- pmount: mount with proper uid and gid options

0.0.2
-----
- pmount: added check if really suid, print an error if not
- pmount: set the real user id to root before executing mount (mount
  refuses to mount a device if real uid != root)

0.0.1
-----
- initial release


pumount.c
/**
 * pumount.c - policy wrapper around 'umount' to allow mounting removable
 *             devices for normal users
 *
 * Author: Martin Pitt <martin.pitt@canonical.com>
 * (c) 2004 Canonical Ltd.
 *
 * This software is distributed under the terms and conditions of the 
 * GNU General Public License. See file GPL for the full text of the license.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/wait.h>
#include <limits.h>
#include <getopt.h>
#include <libintl.h>
#include <sys/stat.h>

#include "policy.h"
#include "utils.h"

/* error codes */
const int E_ARGS = 1;
const int E_DEVICE = 2;
const int E_POLICY = 4;
const int E_EXECUMOUNT = 5;
const int E_INTERNAL = 100;

static char mntpt[MEDIA_STRING_SIZE];

/**
 * Print some help.
 * @param exename Name of the executable (argv[0]).
 */
void
usage( const char* exename )
{
    printf( _("Usage:\n\n%s [options] <device>\n"
    "  Umount <device> from a directory below %s if policy requirements\n"
    "  are met (see pumount(1) for details). The mount point directory is removed\n"
    "  afterwards.\n\n"
    "Options:\n"
    "  -l, --lazy : umount lazily, see umount(8)\n"
    "  -d, --debug : enable debug output (very verbose)\n"
    "  -h, --help  : print help message and exit successfuly\n"),
        exename, MEDIADIR );
}

/**
 * Check whether the user is allowed to umount the given device.
 * @param do_lazy Flag if a lazy unmount is requested (in this case the device
 *        node does not need to exist)
 * @return 0 on success, -1 on failure
 */
int
check_umount_policy( const char* device, int do_lazy ) 
{
    int devvalid;

    devvalid = ( do_lazy || device_valid( device ) ) &&
        device_mounted( device, 1, mntpt );

    if( !devvalid )
        return -1;

    /* paranoid check */
    if( !mntpt || !*mntpt ) {
        fputs( _("Internal error: could not determine mount point\n"), stderr );
        exit( E_INTERNAL );
    }

    /* mount point must be below MEDIADIR */
    if( strncmp( mntpt, MEDIADIR, sizeof( MEDIADIR )-1 ) ) {
        fprintf( stderr, _("Error: mount point %s is not below %s\n"), mntpt,
                MEDIADIR );
        return -1;
    }

    debug( "policy check passed\n" );
    return 0;
}

/**
 * Drop all privileges and exec 'umount device'. Does not return on success, if
 * it returns, UMOUNTPROG could not be executed.
 * @param lazy 0 for normal umount, 1 for lazy umount
 */
void
do_umount_fstab( const char* device, int lazy )
{
    /* drop all privileges */
    get_root();
    if( setuid( getuid() ) ) {
        perror( _("Error: could not drop all uid privileges") );
        return;
    }

    debug( "device %s handled by fstab, calling umount\n", device );

    if( lazy )
        execl( UMOUNTPROG, UMOUNTPROG, "-l", device, NULL );
    else
        execl( UMOUNTPROG, UMOUNTPROG, device, NULL );
    perror( _("Error: could not execute umount") );
}

/**
 * Raise to full privileges and mounts device to mntpt.
 * @param lazy 0 for normal umount, 1 for lazy umount
 * @return 0 on success, -1 if UMOUNTPROG could not be executed.
 */
int
do_umount( const char* device, int do_lazy )
{
    int status;

    if( do_lazy )
        status = spawn( SPAWN_EROOT|SPAWN_RROOT, UMOUNTPROG, UMOUNTPROG, "-l",
                device, NULL );
    else
        status = spawn( SPAWN_EROOT|SPAWN_RROOT, UMOUNTPROG, UMOUNTPROG,
                device, NULL );

    if( status != 0 ) {
        fprintf( stderr, _("Error: umount failed\n") );
        return -1;
    }

    return 0;
}

/**
 * Check whether device is mapped through cryptsetup, and release it if so.
 */
void
release_encrypted( const char* device )
{
    spawn( SPAWN_EROOT|SPAWN_NO_STDOUT|SPAWN_NO_STDERR, CRYPTSETUP, CRYPTSETUP,
            "luksClose", device, NULL );
}

/**
 * Entry point.
 */
int
main( int argc, char** argv )
{
    char device[PATH_MAX], mntptdev[PATH_MAX], path[PATH_MAX];
    const char* fstab_device;
    char* dmlabel;
    int is_real_path = 0;
    int do_lazy = 0;
    struct stat st;

    int  option;
    static struct option long_opts[] = {
        { "help", 0, NULL, 'h'},
        { "debug", 0, NULL, 'd'},
        { "lazy", 0, NULL, 'l'},
        { NULL, 0, NULL, 0}
    };

    /* initialize locale */
    setlocale( LC_ALL, "" );
    bindtextdomain( "pmount", NULL );
    textdomain( "pmount" );

    /* are we root? */
    if( geteuid() ) {
        fputs( _("Error: this program needs to be installed suid root\n"), stderr );
        return E_INTERNAL;
    }

    /* drop root privileges until we really need them (still available as saved uid) */
    seteuid( getuid() );

    /* parse command line options */
    do {
        switch( option = getopt_long( argc, argv, "+hdlu", long_opts, NULL ) ) {
            case -1:        break;          /* end of arguments */
            case '?':        return E_ARGS;  /* unknown argument */

            case 'h':        usage( argv[0] ); return 0;

            case 'd':   enable_debug = 1; break;

            case 'l':        do_lazy = 1; break;

            default:
                fprintf( stderr, _("Internal error: getopt_long() returned unknown value\n") );
                return E_INTERNAL;
        }
    } while( option != -1 );

    /* invalid number of args? */
    if( optind + 1 != argc ) {
        usage( argv[0] );
        return E_ARGS;
    }

    /* if we got a mount point, convert it to a device */
    debug ("checking whether %s is a mounted directory\n", argv[optind]);
    if( fstab_has_mntpt( "/proc/mounts", argv[optind], mntptdev, sizeof(mntptdev) ) ) {
        debug( "resolved mount point %s to device %s\n", argv[optind], mntptdev );
        argv[optind] = mntptdev;
    } else if( !strchr( argv[optind], '/' ) ) {
        /* try to prepend MEDIADIR */
        snprintf( path, sizeof( path ), "%s%s", MEDIADIR, argv[optind] );
        debug ("checking whether %s is a mounted directory\n", path);
        if( fstab_has_mntpt( "/proc/mounts", path, mntptdev, sizeof(mntptdev) ) ) {
            debug( "resolved mount point %s to device %s\n", path, mntptdev );
            argv[optind] = mntptdev;
        }
    }

    /* get real path, if possible */
    if( realpath( argv[optind], device ) ) {
        debug( "resolved %s to device %s\n", argv[optind], device );
        is_real_path = 1;
    } else {
        debug( "%s cannot be resolved to a proper device node\n", argv[optind] );
        snprintf( device, sizeof( device ), "%s", argv[optind] );
    }

    /* is the device already handled by fstab? */
    fstab_device = fstab_has_device( "/etc/fstab", device, NULL, NULL );
    if( fstab_device ) {
        do_umount_fstab( fstab_device, do_lazy );
        return E_EXECUMOUNT;
    }

    /* we cannot really check the real path when unmounting lazily since the
     * device node might not exist any more */
    if( !is_real_path && !do_lazy ) {
        perror( _("Error: could not determine real path of the device") );
        return E_DEVICE;
    }

    /* does the device start with DEVDIR? */
    if( strncmp( device, DEVDIR, sizeof( DEVDIR )-1 ) ) {
        fprintf( stderr, _("Error: invalid device %s (must be in /dev/)\n"), device );
        return E_DEVICE;
    }

    /* check if we have a dmcrypt device */
    dmlabel = strreplace( device, '/', '_' );
    snprintf( path, sizeof( path ), "/dev/mapper/%s", dmlabel );
    free( dmlabel );
    if( !stat( path, &st ) ) {
        snprintf( device, sizeof( device ), "%s", path );
        debug( "Unmounting mapped device %s instead.\n", device );
    }

    if( check_umount_policy( device, do_lazy ) )
        return E_POLICY;

    /* go for it */
    if( do_umount( device, do_lazy ) )
        return E_EXECUMOUNT;

    /* release LUKS device, if appropriate */
    release_encrypted( device );

    /* delete mount point */
    remove_pmount_mntpt( mntpt );

    return 0; 
}
